Document new upload mechanism (#835)
mozilla/glean: docs/dev/ffi/when-to-use-what-in-the-ffi.md

### Primitives

Numeric primitives are the easiest to pass between languages.
The main recommendation is: use the equivalent and same-sized type as the one provided by Rust.

There are a couple of exceptions/caveats, especially for Kotlin. All of them are caused by JNA/Android issues.
Swift has very good support for calling over the FFI.

1. `bool`: Don't use it. JNA doesn't handle it well. Instead, use a numeric type
    (like `u8`) and represent 0 for `false` and 1 for `true` for interchange over the
    FFI, converting back to Kotlin's `Boolean` or Swift's `Bool` after (as to
    not expose this somewhat annoying limitation in our public API).
    All wrappers already include utility functions to turn 8-bit integers (`u8`) back to booleans
    (`toBool()` or equivalent methods).

2. `usize`/`isize`: These cause the structure size to be different based on the
   platform. JNA does handle this if you use `NativeSize`, but it's awkward.
   Use the size-defined integers instead, such as `i64`/`i32` and their language-equivalents
   (Kotlin: `Long`/`Int`, Swift:`UInt64`/`UInt32`).

>    *Caution:* In Kotlin integers are signed by default. You can use `u64`/`u32` for `Long`/`Int`, if you're certain the value is not negative.

Reviewer: Ugh. I don't like "if you're certain". Programmers make mistakes (or maybe that's just me) so I'd prefer some sort of "You _can_ do this if you ensure the values are non-negative through asserts or error reporting code" or something

Author: will change. I do agree, but for now that's the state we're in :/

Author: ```suggestion
   *Caution:* In Kotlin integers are signed by default. You can use `u64`/`u32` for `Long`/`Int` if you ensure the values are non-negative through asserts or error reporting code.
```

    to the destructor, it will allocate a temporary buffer, pass it to Rust, which
    we'll free, corrupting both heaps 💥. Oops!

2. If the string is passed into Rust from Kotlin/Swift, the Rust code should
   declare the parameter as a [`FfiStr<'_>`](https://docs.rs/ffi-support/*/ffi_support/struct.FfiStr.html).
   and things should then work more or less automatically. The `FfiStr` has methods
   for extracting it's data as `&str`, `Option<&str>`, `String`, and `Option<String>`.

### Aggregates

This is any type that's more complex than a primitive or a string (arrays,
structures, and combinations there-in).
There are two options we recommend for these cases:

> 1. Passing data as JSON. This is very easy, and useful for prototyping, however
>    much slower, requires a great deal of copying and redundant encode/decode
>    steps (in general, the data will be copied at least 4 times to make this

Reviewer: ```suggestion
1. Passing data as JSON. This is very easy and useful for prototyping, but is
   much slower and requires a great deal of copying and redundant encode/decode
   steps (in general, the data will be copied at least 4 times to make this
```

>    work, and almost certainly more in practice),
>    and can be done relatively easily by `derive(Serialize, Deserialize)`,
>    and converting to a JSON string using `serde_json::to_string`.

Reviewer: ```suggestion
   work, and almost certainly more in practice).
   It can be done relatively easily by `derive(Serialize, Deserialize)`,
   and converting to a JSON string using `serde_json::to_string`.
```

>    This is a viable option for test-only function, where the overhead is not important.

Reviewer: ```suggestion
   This is a viable option for test-only functions, where the overhead is not important.
```
